### PR TITLE
Cuda kernels & Kotekan stages from Dustin's MVP work

### PR DESCRIPTION
**Pull request by dstndstn**
This follows on from #1107 (and includes all commits from #1107 !), adding the Cuda kernels and MVP config files:

- FRB beamformer
- upchannelizer (two sample lengths)
- FRB beam reformer (FRB2 kernel)
- quantizer (FRB3 kernel)
- rechunker (goes between FRB2 and FRB3 kernels)
- chord MVP setup (sets up buffer views to hack our way around some data-length differences)

### CODE REVIEW

**Comment by andrerenard** [comment]
@dstndstn looks like merging 1107 has created some conflicts with the base branch.

**Comment by dstndstn** [comment]
Yup, I was just looking at that :)

**Comment by dstndstn** [comment]
Conflicts resolved and CI tests passing aside from the usual 18.04 numerical failure

File: lib/cuda/chordMVPSetup.cpp
Comment (andrerenard) [inline lib/cuda/chordMVPSetup.cpp:8]: I think we can just remove this class based on the discussion yesterday.  Let me know if you want to make the change in this PR, or wait until we merge everything in, and then push the new model on top of this?

File: lib/cuda/chordMVPSetup.cpp
Comment (andrerenard) [inline lib/cuda/chordMVPSetup.cpp]: Comments like this are better just before the start of the class declaration, using doxygen style comments.  This way we can automatically pull the comments into the readthedocs documentation.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.cpp]: FWIW.  I like `TODO` instead of `FIXME/HACK`, since a lot of IDEs will automatically find those for you.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.cpp]: These feel like they could also become util functions - along with simple doxygen strings.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.cpp:17]: What does this do?  Could you add a more complete comment.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.cpp:40]: Does (or could) that beamforming note live on the CHORD wiki? If so could we link it?  I could see someone trying to find that later down the road.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.cpp]: Remove commented code - unless it includes a comment about why it's commented - and is useful to future users.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.cpp]: Remove if not being used?

File: lib/cuda/cudaFRBBeamReformer.hpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamReformer.hpp:18]: Might be worth adding that this contains a lot of MVP level logic for the phases to this comment as well, since this is what will be pulled into the docs.

File: lib/cuda/cudaFRBBeamformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaFRBBeamformer.cpp:266]: I think we want to put this on the output data copy stream.  Otherwise if the previous pipeline is still copying out the data, then this will stall the kernel pipeline.

File: lib/cuda/cudaQuantize.cpp
Comment (andrerenard) [inline lib/cuda/cudaQuantize.cpp]: Remove?

File: lib/cuda/cudaQuantize.hpp
Comment (andrerenard) [inline lib/cuda/cudaQuantize.hpp:38]: Maybe add a comment explaining where these come from?

File: lib/cuda/cudaUpchannelize.cpp
Comment (andrerenard) [inline lib/cuda/cudaUpchannelize.cpp:73]: Bit of an aside question, but would we ever want to change this value in the config?  Say we target a newer GPU?

File: lib/cuda/cudaUpchannelize.cpp
Comment (andrerenard) [inline lib/cuda/cudaUpchannelize.cpp:175]: Same concern here about pipeline stalls.

File: lib/testing/gpuSimulateCudaFRBBeamformer.cpp
Comment (andrerenard) [inline lib/testing/gpuSimulateCudaFRBBeamformer.cpp]: Should this be WARN?  Generate an error, or keep going as it does?

File: lib/testing/gpuSimulateCudaFRBBeamformer.cpp
Comment (andrerenard) [inline lib/testing/gpuSimulateCudaFRBBeamformer.cpp]: ?

File: lib/testing/gpuSimulateCudaFRBBeamformer.cpp
Comment (andrerenard) [inline lib/testing/gpuSimulateCudaFRBBeamformer.cpp]: These look generically useful.

File: lib/testing/gpuSimulateCudaUpchannelize.cpp
Comment (andrerenard) [inline lib/testing/gpuSimulateCudaUpchannelize.cpp]: Is there a cleaner way to do this?  perhaps with a config option, or just remove the else clause?

File: lib/testing/gpuSimulateCudaUpchannelize.cpp
Comment (andrerenard) [inline lib/testing/gpuSimulateCudaUpchannelize.cpp:39]: Will the code even run if `!KOTEKAN_FLOAT16`?  Should this entire file depend on support for that?

**Review by andrerenard — CHANGES_REQUESTED** [review]
Overall looks really good.  I'm not mentioning any changes that relate to the redesign we talked about.  A few comments inline.

File: lib/cuda/chordMVPSetup.cpp
Comment (dstndstn) [reply lib/cuda/chordMVPSetup.cpp:8]: I'd like to keep this for now, so that the MVP continues to work -- and then the PR that removes this and replaces it with the new thing will make more sense and have greater glory!

File: lib/cuda/chordMVPSetup.cpp
Comment (dstndstn) [reply lib/cuda/chordMVPSetup.cpp]: Done

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.cpp]: Done

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.cpp]: Done

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.cpp:17]: I punted by referring to the equation number in Kendrick's notes :)

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.cpp]: Done

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.cpp]: Yes, thanks, done

File: lib/cuda/cudaFRBBeamReformer.hpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.hpp:18]: Good point, done

File: lib/cuda/cudaQuantize.cpp
Comment (dstndstn) [reply lib/cuda/cudaQuantize.cpp]: Yes, done!

File: lib/cuda/cudaQuantize.hpp
Comment (dstndstn) [reply lib/cuda/cudaQuantize.hpp:38]: Done

File: lib/cuda/cudaUpchannelize.cpp
Comment (dstndstn) [reply lib/cuda/cudaUpchannelize.cpp:73]: That's a good idea!  Will consider this in Erik's auto-generated kotekan wrappers!

File: lib/testing/gpuSimulateCudaFRBBeamformer.cpp
Comment (dstndstn) [reply lib/testing/gpuSimulateCudaFRBBeamformer.cpp]: Sure, I switched to a WARN.  Since this is only for testing, maybe it should be an ERROR.

File: lib/testing/gpuSimulateCudaFRBBeamformer.cpp
Comment (dstndstn) [reply lib/testing/gpuSimulateCudaFRBBeamformer.cpp]: Removed this ... this was for doing a one-hot phase array, but for the FRB beamformer, that doesn't really work, because it has to compute the full "beamgrid" array which then gets interpolated to the actual beam locations.

File: lib/testing/gpuSimulateCudaFRBBeamformer.cpp
Comment (dstndstn) [reply lib/testing/gpuSimulateCudaFRBBeamformer.cpp]: Moved to visUtil.hpp!

File: lib/testing/gpuSimulateCudaUpchannelize.cpp
Comment (dstndstn) [reply lib/testing/gpuSimulateCudaUpchannelize.cpp]: Sure, I just removed the `else` clause.

File: lib/testing/gpuSimulateCudaUpchannelize.cpp
Comment (dstndstn) [reply lib/testing/gpuSimulateCudaUpchannelize.cpp:39]: Right, if there's no FLOAT16, it raises a WARN in the main_thread().

File: lib/cuda/cudaFRBBeamformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamformer.cpp:266]: We decided to make this a debug-mode-only thing, which we'll do in Erik's boilerplate factory.

File: lib/cuda/cudaFRBBeamReformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaFRBBeamReformer.cpp:40]: Yes, this is a good idea!  I will punt to Kendrick

File: lib/cuda/cudaFRBBeamformer.cpp
Comment (andrerenard) [reply lib/cuda/cudaFRBBeamformer.cpp:266]: Also remember we need to `mlock` and register the memory we are using here with CUDA (unless it's being created with `cudaMalloc`, then you just need to check that it's getting locked).

File: config/tests/rechunk.yaml
Comment (andrerenard) [inline config/tests/rechunk.yaml]: Missing comment.

File: config/tests/rechunk.yaml
Comment (andrerenard) [inline config/tests/rechunk.yaml]: This test only prints things, it doesn't test them, should we keep the config in the repo, or was this intended as a temporary debugging config?

**Review by andrerenard — APPROVED** [review]

**Pull request merged**
